RFC: Fix LibGit2.rebase! (Fixes #16366)

kmsquire opened the pull request:
Right now, `LibGit2.rebase!` gives

``` julia
julia> LibGit2.rebase!(repo, "master")
ERROR: GitError(Code:ERROR, Class:Invalid, Invalid version 0 on git_checkout_options)
 [inlined code] from ./libgit2/error.jl:98
 in #GitRebase#73(::Nullable{Base.LibGit2.GitAnnotated}, ::Base.LibGit2.RebaseOptions, ::Type{T}, ::Base.LibGit2.GitRepo, ::Base.LibGit2.GitAnnotated, ::Base.LibGit2.GitAnnotated) at ./libgit2/rebase.jl:7
 in (::Base.LibGit2.##120#122{Base.LibGit2.GitRepo,String})(::Base.LibGit2.GitReference) at ./libgit2.jl:429
 in with(::Base.LibGit2.##120#122{Base.LibGit2.GitRepo,String}, ::Base.LibGit2.GitReference) at ./libgit2/types.jl:613
 in rebase!(::Base.LibGit2.GitRepo, ::String) at ./libgit2.jl:413
 in eval(::Module, ::Any) at ./boot.jl:226
```

[`checkout_opts::CheckoutOptions` is the last member of `RebaseOptions`](https://github.com/JuliaLang/julia/blob/5d52f0273d3ba8130da22af52083a71924b0c059/base/libgit2/types.jl#L450-L456).  

Digging into this, I found that [`MergeOptions`](https://github.com/JuliaLang/julia/blob/5d52f0273d3ba8130da22af52083a71924b0c059/base/libgit2/types.jl#L360-L368) seems to be missing a couple of members from [`git_merge_options`](https://github.com/libgit2/libgit2/blob/c148533024319689fca016f7c452d556265bb13f/include/git2/merge.h#L238-L290), which would cause a misalignment of the members of `RebaseOptions` when passed to libgit2.

Unfortunately, the error is still there even after this change, so either I didn't do this right, or there seems to be some other reason for the error.

I don't know when I'll have time to work on this again, so any help in finishing it up would be appreciated.

Cc: @nalimilan @wildart @tkelman

kmsquire committed:
Add missing members to LibGit2.MergeOptions

* These seem to have been missed in #15587
* Note that this PR includes a commented out field, default_driver,
  which does not exist in 0.24.1, but does appear on the current
  libgit2 master

nalimilan commented:
The error you get now on CI is different:
```
LoadError: GitError(Code:ENOTFOUND, Class:Config, Config value 'user.name' was not found)
```

So there must be another bug elsewhere. We could merge this without the tests, as it's an improvement anyways.

kmsquire commented:
I'm still getting the other error locally, on OSX.

Ironically, travis for OSX passed, but that's only because [`libgit2` is version 0.23.4 there](https://travis-ci.org/JuliaLang/julia/jobs/130517276#L482-L497), and the bug happens on 0.24.

I wonder if some of the assumed type sizes of things like `git_merge_flag_t`, `git_merge_file_favor_t`, etc. are different on OSX...

wildart commented:
No, error relates to `LibGit2.rebase!`. I've seen it before with travis. There is no git configuration in travis test system, so `LibGit2.default_signature` throws error when could not get certain properties from the configuration. I think some dummy git config should set (user & e-mail) in travis system before calling these tests.

kmsquire commented:
That makes sense. That won't fix my local error, but it should get things
passing here.

On Monday, May 16, 2016, Art <notifications@github.com> wrote:

> No, error relates to LibGit2.rebase!. I've seen it before with travis.
> There is no git configuration in travis test system, so
> LibGit2.default_signature throws error when could not get certain
> properties from the configuration. I think some dummy git config should set
> (user & e-mail) in travis system before calling these tests.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/JuliaLang/julia/pull/16382#issuecomment-219478357>
>

tkelman commented:
We used to do that, but got rid of it when libgit2 made it no longer necessary. Why would we start needing it again?

tkelman commented on test/libgit2.jl:
```diff
@@ -340,6 +340,11 @@ mktempdir() do dir
+            # Set the username and email for the test_repo
+            cfg = LibGit2.GitConfig(repo)
+            LibGit2.set!(cfg, "user.name", "AAAA")
+            LibGit2.set!(cfg, "user.email", "BBBB@BBBB.COM")
```
would it maybe make sense to wait to set this until as late as possible, only for the part of the test that needs it?

kmsquire replied:
Okay.

kmsquire committed:
Added test for LibGit2.rebase!

kmsquire commented:
Well, adding `user.name` and `user.email` at least brought the correct error forward (on AppVeyor; travis hasn't run yet).

It turns out that one of the fields I had added, `default_driver`, appears on master, but not on `v0.24.x`.

tkelman commented on base/libgit2/types.jl:
```diff
@@ -357,28 +357,60 @@ immutable DiffDelta
+        target_limit::Cuint
+        metric::Ptr{Void}
+        recursion_limit::Cuint
+        # default_driver::Cstring  # TODO: added after v"0.24.1"
```
presumably they won't backport a new field in a point release? so 0.25.0 will be the first to have this?

tkelman commented:
Would the test here have caught the problem from #16370? Might also be good to add a keyword for `inmemory` there.

kmsquire committed:
Add speculative MergeOptions change for libgit2 v0.25.0

* on libgit2 master, git_merge_options has an additional field,
  default_driver.  Presumably this modification will be released
  with v0.25.0, so test for that now.

kmsquire committed:
Add inmemory keyword option to RebaseOptions for libgit2 v0.24.0

* #16370 set this option to a constant value

kmsquire commented:
> Would the test here have caught the problem from #16370? Might also be good to add a keyword for `inmemory` there.

I was just adding that keyword.  Yes, the tests raises an error for that problem if I test it before #16370 was merged.

kmsquire commented:
Quick question: I added b21e83c to handle the new field in `git_merge_options` on the libgit2 master branch, but it's not going to work until libgit2 bumps their version (and there's always the possibility of yet another change to this struct).

Should I keep or remove that commit?

tkelman commented:
I say keep it. It will either be the only change needed to this struct when 0.25.0 comes out in which case it's work that's been done early, or we'll need to make more changes. I don't think it hurts anything to have it maybe-right in preparation vs not there.

kmsquire commented:
Just to verify, this is the current test on baf68d1, the commit before #16370:

```julia
$ make test-libgit2
    JULIA test/libgit2
     * libgit2              ERROR: LoadError: LoadError: MethodError: no method matching Base.LibGit2.RebaseOptions(::UInt32, ::Int32, ::Cstring, ::Base.LibGit2.MergeOptions, ::Base.LibGit2.CheckoutOptions)
Closest candidates are:
  Base.LibGit2.RebaseOptions(::Any, ::Any, ::Any, ::Any, ::Any, !Matched::Any)
  Base.LibGit2.RebaseOptions(::UInt32, ::Int32, !Matched::Int32, !Matched::Cstring, !Matched::Base.LibGit2.MergeOptions, !Matched::Base.LibGit2.CheckoutOptions)
  Base.LibGit2.RebaseOptions{T}(::Any)
 in (::Base.LibGit2.##119#121{Base.LibGit2.GitRepo,String})(::Base.LibGit2.GitReference) at ./libgit2.jl:429
 in with(::Base.LibGit2.##119#121{Base.LibGit2.GitRepo,String}, ::Base.LibGit2.GitReference) at ./libgit2/types.jl:581
 [inlined code] from ./libgit2.jl:413
 in (::##29#40)(::String) at /Users/kevin/Source/julia/test/libgit2.jl:369
 in mktempdir(::##29#40, ::String) at ./file.jl:272
 [inlined code] from ./tuple.jl:8
 in mktempdir(::Function) at ./file.jl:270
 in include_from_node1(::String) at ./loading.jl:426
 [inlined code] from ./util.jl:179
 in runtests(::String) at /Users/kevin/Source/julia/test/testdefs.jl:7
 [inlined code] from ./promotion.jl:248
 in (::Base.##279#280{##16#24,Tuple{String},Array{Any,1}})() at ./multi.jl:781
 in run_work_thunk(::Base.##279#280{##16#24,Tuple{String},Array{Any,1}}, ::Bool) at ./multi.jl:754
 in #remotecall_fetch#284(::Array{Any,1}, ::Function, ::Function, ::LocalProcess, ::String, ::Vararg{String,N}) at ./multi.jl:801
 [inlined code] from ./promotion.jl:248
 in #remotecall_fetch#286(::Array{Any,1}, ::Function, ::Function, ::Int64, ::String, ::Vararg{String,N}) at ./multi.jl:817
 in remotecall_fetch(::Function, ::Int64, ::String, ::Vararg{String,N}) at /Users/kevin/Source/julia/usr/lib/julia/sys.dylib:-1
 [inlined code] from /Users/kevin/Source/julia/test/runtests.jl:36
 in (::##15#23{Array{Any,1}})() at ./task.jl:308
while loading /Users/kevin/Source/julia/test/libgit2.jl, in expression starting on line 68
 in #remotecall_fetch#284(::Array{Any,1}, ::Function, ::Function, ::LocalProcess, ::String, ::Vararg{String,N}) at ./multi.jl:802
 [inlined code] from ./promotion.jl:248
 in #remotecall_fetch#286(::Array{Any,1}, ::Function, ::Function, ::Int64, ::String, ::Vararg{String,N}) at ./multi.jl:817
 in remotecall_fetch(::Function, ::Int64, ::String, ::Vararg{String,N}) at /Users/kevin/Source/julia/usr/lib/julia/sys.dylib:-1
 [inlined code] from /Users/kevin/Source/julia/test/runtests.jl:36
 in (::##15#23{Array{Any,1}})() at ./task.jl:308
 in sync_end() at ./task.jl:274
 [inlined code] from ./task.jl:283
 in (::##11#19)() at /Users/kevin/Source/julia/test/runtests.jl:31
 in cd(::##11#19, ::String) at ./file.jl:48
 in include_from_node1(::String) at ./loading.jl:426
 in process_options(::Base.JLOptions) at ./client.jl:263
 in _start() at ./client.jl:319
while loading /Users/kevin/Source/julia/test/runtests.jl, in expression starting on line 13
Makefile:12: recipe for target 'libgit2' failed
make[1]: *** [libgit2] Error 1
Makefile:575: recipe for target 'test-libgit2' failed
make: *** [test-libgit2] Error 2
```

tkelman commented:
The error in https://ci.appveyor.com/project/JuliaLang/julia/build/1.0.1879/job/qs0pq3mc0n5rq9bt looks rather scary and there's a bit in there that says `Base.InitError(mod=:LibGit2, error=ErrorException("could not load library "libgit2"
The paging file is too small for this operation to complete.` but I'd be a bit surprised if this PR were the cause. Restarting it just in case.
